qlever.py

```diff
@@ -0,0 +1,488 @@
+#!/usr/bin/env python3
+
+# This is the QLever script (new version, written in Python).
+
+from configparser import ConfigParser, ExtendedInterpolation
+from datetime import datetime, date
+import os
+import glob
+import psutil
+import re
+import shlex
+import subprocess
+import sys
+import time
+import traceback
+
+BLUE = "\033[34m"
+RED = "\033[31m"
+BOLD = "\033[1m"
+NORMAL = "\033[0m"
+
+
+class ActionException(Exception):
+    pass
+
+
+class Actions:
+
+    def __init__(self):
+        self.config = ConfigParser(interpolation=ExtendedInterpolation())
+        self.config.read("Qleverfile.ini")
+        self.name = self.config['DEFAULT']['name']
+        self.yes_values = ["1", "true", "yes"]
+
+        # Default values for options that are not mandatory in the Qleverfile.
+        defaults = {
+            "DEFAULT": {
+                "log-level": "info",
+            },
+            "server": {
+                "binary": "ServerMain",
+                "num_threads": "8",
+                "cache_max_size_gb": "5",
+                "cache_max_size_gb_single_entry": "1",
+                "cache_max_num_entries": "100",
+                "with_text_index": "no",
+                "only_pso_and_pos_permutations": "no",
+                "no_patterns": "no",
+            },
+            "index": {
+                "binary": "IndexBuilderMain",
+                "with_text_index": "no",
+                "only_pso_and_pos_permutations": "no",
+                "no_patterns": "no",
+            },
+            "docker": {
+                "image": "adfreiburg/qlever",
+                "container_server": f"qlever.server.{self.name}",
+                "container_indexer": f"qlever.indexer.{self.name}",
+            },
+            "ui": {
+                "port": "7000",
+            }
+        }
+        for section in defaults:
+            for option in defaults[section]:
+                if not self.config[section].get(option):
+                    self.config[section][option] = defaults[section][option]
+
```

Comment:
Is there also a check, that all the configurations from the Qleverfile are also valid options?
For example what happens if I use `num_treads` in the QLeverfile or a similar typo? I think rasing a hard error would be the most user-friendly, but therefore we need the set of all supported options as a whitelist (similar to the defaults).